File: LeonEdBackend/Services/Implementations/SchoolService.cs
using Microsoft.EntityFrameworkCore;
using LeonEdBackend.Data;
using LeonEdBackend.DTOs.Common;
using LeonEdBackend.DTOs.SchoolDtos;
using LeonEdBackend.Helpers;
using LeonEdBackend.Services.Interfaces;

namespace LeonEdBackend.Services.Implementations
{
    public class SchoolService : ISchoolService
    {
        private readonly AppDbContext _context;

        public SchoolService(AppDbContext context)
        {
            _context = context;
        }

        public async Task<ApiResponse<PagedResult<SchoolResponse>>> GetAllSchools(PaginationParams paginationParams)
        {
            var query = _context.Schools.AsQueryable();

            if (!string.IsNullOrWhiteSpace(paginationParams.Search))
            {
                var search = paginationParams.Search.ToLower();
                query = query.Where(s => s.Name.ToLower().Contains(search) ||
                                         s.ContactEmail.ToLower().Contains(search));
            }

            var totalCount = await query.CountAsync();

            var schools = await query
                .OrderByDescending(s => s.CreatedAt)
                .Skip((paginationParams.PageNumber - 1) * paginationParams.PageSize)
                .Take(paginationParams.PageSize)
                .Select(s => new SchoolResponse
                {
                    Id = s.Id,
                    Name = s.Name,
                    Address = s.Address,
                    ContactEmail = s.ContactEmail,
                    ContactPhone = s.ContactPhone,
                    LogoUrl = s.LogoUrl,
                    Slug = s.Slug,
                    SubscriptionPlan = s.SubscriptionPlan.ToString(),
                    SubscriptionStatus = s.SubscriptionStatus.ToString(),
                    IsActive = s.IsActive,
                    MaxTeachers = s.SubscriptionPlan == SubscriptionPlan.Free ? 20 :
                                  s.SubscriptionPlan == SubscriptionPlan.Plus ? 30 : int.MaxValue,
                    MaxStudents = s.SubscriptionPlan == SubscriptionPlan.Free ? 100 :
                                  s.SubscriptionPlan == SubscriptionPlan.Plus ? 200 : int.MaxValue,
                    CurrentTeacherCount = s.Teachers.Count(t => t.IsActive),
                    CurrentStudentCount = s.Students.Count(st => st.Status == StudentStatus.Active),
                    CreatedAt = s.CreatedAt
                })
                .ToListAsync();

            var result = new PagedResult<SchoolResponse>
            {
                Items = schools,
                TotalCount = totalCount,
                PageNumber = paginationParams.PageNumber,
                PageSize = paginationParams.PageSize
            };

            return ApiResponse<PagedResult<SchoolResponse>>.SuccessResponse(result);
        }

        public async Task<ApiResponse<SchoolResponse>> GetSchoolById(Guid id)
        {
            var school = await _context.Schools
                .Include(s => s.Teachers)
                .Include(s => s.Students)
                .FirstOrDefaultAsync(s => s.Id == id);

            if (school == null)
            {
                return ApiResponse<SchoolResponse>.FailResponse("School not found.");
            }

            var response = MapToResponse(school);
            return ApiResponse<SchoolResponse>.SuccessResponse(response);
        }

        public async Task<ApiResponse<SchoolResponse>> UpdateSchool(Guid id, UpdateSchoolRequest request)
        {
            var school = await _context.Schools
                .Include(s => s.Teachers)
                .Include(s => s.Students)
                .FirstOrDefaultAsync(s => s.Id == id);

            if (school == null)
            {
                return ApiResponse<SchoolResponse>.FailResponse("School not found.");
            }

            if (!string.IsNullOrWhiteSpace(request.Name)) school.Name = request.Name;
            if (!string.IsNullOrWhiteSpace(request.Address)) school.Address = request.Address;
            if (!string.IsNullOrWhiteSpace(request.ContactEmail)) school.ContactEmail = request.ContactEmail;
            if (!string.IsNullOrWhiteSpace(request.ContactPhone)) school.ContactPhone = request.ContactPhone;
            if (!string.IsNullOrWhiteSpace(request.LogoUrl)) school.LogoUrl = request.LogoUrl;
            school.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            var response = MapToResponse(school);
            return ApiResponse<SchoolResponse>.SuccessResponse(response, "School updated successfully.");
        }

        public async Task<ApiResponse<SchoolResponse>> UpdateSchoolPlan(Guid schoolId, SubscriptionPlan plan)
        {
            var school = await _context.Schools
                .Include(s => s.Teachers)
                .Include(s => s.Students)
                .FirstOrDefaultAsync(s => s.Id == schoolId);

            if (school == null)
            {
                return ApiResponse<SchoolResponse>.FailResponse("School not found.");
            }

            school.SubscriptionPlan = plan;
            school.UpdatedAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();

            var response = MapToResponse(school);
            return ApiResponse<SchoolResponse>.SuccessResponse(response, "Subscription plan updated successfully.");
        }

        public async Task<ApiResponse<bool>> UpdateSchoolStatus(Guid schoolId, bool isActive)
        {
            var school = await _context.Schools.FindAsync(schoolId);
            if (school == null)
            {
                return ApiResponse<bool>.FailResponse("School not found.");
            }

            school.IsActive = isActive;
            school.SubscriptionStatus = isActive ? SubscriptionStatus.Active : SubscriptionStatus.Suspended;
            school.UpdatedAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();

            var message = isActive ? "School activated successfully." : "School suspended successfully.";
            return ApiResponse<bool>.SuccessResponse(true, message);
        }

        public Task<List<SubscriptionPlanInfo>> GetSubscriptionPlans()
        {
            var plans = new List<SubscriptionPlanInfo>
            {
                new() { Name = "Free", MaxTeachers = 20, MaxStudents = 100, Description = "Basic plan with up to 20 teachers and 100 students." },
                new() { Name = "Plus", MaxTeachers = 30, MaxStudents = 200, Description = "Enhanced plan with up to 30 teachers and 200 students." },
                new() { Name = "Premium", MaxTeachers = int.MaxValue, MaxStudents = int.MaxValue, Description = "Unlimited teachers and students." }
            };
            return Task.FromResult(plans);
        }

        private static SchoolResponse MapToResponse(Models.School school)
        {
            return new SchoolResponse
            {
                Id = school.Id,
                Name = school.Name,
                Address = school.Address,
                ContactEmail = school.ContactEmail,
                ContactPhone = school.ContactPhone,
                LogoUrl = school.LogoUrl,
                Slug = school.Slug,
                SubscriptionPlan = school.SubscriptionPlan.ToString(),
                SubscriptionStatus = school.SubscriptionStatus.ToString(),
                IsActive = school.IsActive,
                MaxTeachers = school.MaxTeachers,
                MaxStudents = school.MaxStudents,
                CurrentTeacherCount = school.Teachers.Count(t => t.IsActive),
                CurrentStudentCount = school.Students.Count(s => s.Status == StudentStatus.Active),
                CreatedAt = school.CreatedAt
            };
        }
    }
}

File: src/services/school.service.ts
import { prisma } from "../config/db";
import { successResponse, failResponse, createPagedResult } from "../utils/response";

export class SchoolService {
  private static mapToResponse(school: any) {
    const currentTeacherCount = school.teachers ? school.teachers.filter((t: any) => t.isActive).length : 0;
    const currentStudentCount = school.students ? school.students.filter((s: any) => s.status === "Active").length : 0;

    return {
      id: school.id,
      name: school.name,
      address: school.address,
      contactEmail: school.contactEmail,
      contactPhone: school.contactPhone,
      logoUrl: school.logoUrl,
      schoolTheme: school.schoolTheme,
      slug: school.slug,
      subscriptionPlan: school.plan?.name || "Free",
      subscriptionStatus: school.subscriptionStatus,
      subscriptionEndedAt: school.subscriptionEndedAt,
      isActive: school.isActive,
      maxTeachers: school.plan?.maxTeachers ?? 20,
      maxStudents: school.plan?.maxStudents ?? 100,
      currentTeacherCount,
      currentStudentCount,
      createdAt: school.createdAt,
    };
  }

  static async getAllSchools(params: any) {
    const pageNumber = parseInt(params.pageNumber || "1", 10);
    const pageSize = parseInt(params.pageSize || "20", 10);
    const search = params.search ? params.search.toLowerCase() : "";

    const where: any = {};
    if (search) {
      where.OR = [
        { name: { contains: search, mode: "insensitive" } },
        { contactEmail: { contains: search, mode: "insensitive" } },
      ];
    }

    const totalCount = await prisma.school.count({ where });

    const schools = await prisma.school.findMany({
      where,
      orderBy: { createdAt: "desc" },
      skip: (pageNumber - 1) * pageSize,
      take: pageSize,
      include: {
        teachers: { select: { isActive: true } },
        students: { select: { status: true } },
        plan: true,
      },
    });

    const items = schools.map((s) => this.mapToResponse(s));
    const pagedResult = createPagedResult(items, totalCount, pageNumber, pageSize);

    return successResponse(pagedResult);
  }

  static async getSchoolById(id: string) {
    const school = await prisma.school.findUnique({
      where: { id },
      include: {
        teachers: { select: { isActive: true } },
        students: { select: { status: true } },
        plan: true,
      },
    });

    if (!school) {
      return failResponse("School not found.");
    }

    return successResponse(this.mapToResponse(school));
  }

  static async updateSchool(id: string, request: any) {
    const school = await prisma.school.findUnique({
      where: { id },
      include: {
        teachers: { select: { isActive: true } },
        students: { select: { status: true } },
      },
    });

    if (!school) {
      return failResponse("School not found.");
    }

    const updatedSchool = await prisma.school.update({
      where: { id },
      data: {
        name: request.name !== undefined ? request.name : undefined,
        address: request.address !== undefined ? request.address : undefined,
        contactEmail: request.contactEmail !== undefined ? request.contactEmail : undefined,
        contactPhone: request.contactPhone !== undefined ? request.contactPhone : undefined,
        logoUrl: request.logoUrl !== undefined ? request.logoUrl : undefined,
        schoolTheme: request.schoolTheme !== undefined ? request.schoolTheme : undefined,
      },
      include: {
        teachers: { select: { isActive: true } },
        students: { select: { status: true } },
        plan: true,
      },
    });

    return successResponse(this.mapToResponse(updatedSchool), "School updated successfully.");
  }

  static async updateSchoolPlan(schoolId: string, planId: string) {
    const school = await prisma.school.findUnique({
      where: { id: schoolId },
      include: {
        teachers: { select: { isActive: true } },
        students: { select: { status: true } },
        plan: true,
      },
    });

    if (!school) {
      return failResponse("School not found.");
    }

    const updatedSchool = await prisma.school.update({
      where: { id: schoolId },
      data: { planId },
      include: {
        teachers: { select: { isActive: true } },
        students: { select: { status: true } },
        plan: true,
      },
    });

    return successResponse(this.mapToResponse(updatedSchool), "Subscription plan updated successfully.");
  }

  static async updateSchoolStatus(schoolId: string, isActive: boolean) {
    const school = await prisma.school.findUnique({
      where: { id: schoolId },
    });

    if (!school) {
      return failResponse("School not found.");
    }

    await prisma.school.update({
      where: { id: schoolId },
      data: {
        isActive,
        subscriptionStatus: isActive ? "Active" : "Suspended",
      },
    });

    const message = isActive ? "School activated successfully." : "School suspended successfully.";
    return successResponse(true, message);
  }

  static async getSubscriptionPlans() {
    const plans = await prisma.paymentPlan.findMany({
      where: { isActive: true },
    });

    return plans.map((p) => ({
      id: p.id,
      name: p.name,
      amount: p.amount,
      maxTeachers: p.maxTeachers,
      maxStudents: p.maxStudents,
      description: `${p.name} plan supporting up to ${
        p.maxTeachers >= 999999 ? "unlimited" : p.maxTeachers
      } teachers and ${p.maxStudents >= 999999 ? "unlimited" : p.maxStudents} students.`,
    }));
  }
}
